### PR TITLE
Fix: Instead of failing, raise a warning if stream selection is performed before config is present

### PR DESCRIPTION
**Pull request by Suraj-Vishwakarma70**
This fixes airbytehq/PyAirbyte#51
@aaronsteers ... Let me know if that doesnt work!

## Summary by CodeRabbit

- **New Features**
  - Introduced logging messages to enhance visibility of stream selection and availability issues.
  - Improved handling for cases with missing configurations or no available streams.

- **Improvements**
  - Enhanced configuration management and stream selection to address pending and incremental stream scenarios.

### CODE REVIEW

**Comment by aaronsteers** [comment]
@coderabbitai review

**Comment by Suraj-Vishwakarma70** [comment]
@aaronsteers I've made one more commit. I got a better approach to solve this! I gues you'll need to run the tests again. Thanks

File: airbyte/sources/base.py [hunk 1]
```diff
@@ -217,10 +222,14 @@ def validate_config(self, config: dict[str, Any] | None = None) -> None:
             )
             raise validation_ex from ex
 
-    def get_available_streams(self) -> list[str]:
+    def get_available_streams(self) -> list[str] | None:
         """Get the available streams from the spec."""
+        if not self._config and self.requires_config():
```
Comment (aaronsteers) [inline airbyte/sources/base.py]: @Suraj-Vishwakarma70 - I might be missing something but I don't see `requires_config()` declared in the code. Is it possible there's another change that didn't get pushed?

**Review by aaronsteers — CHANGES_REQUESTED** [review]

File: airbyte/sources/base.py (hunk 1 above)
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py]: So sorry about that @aaronsteers . 🙇 
I had made a couple of changes after that but it slipped out of my mind to push them.
I've pushed them and you can review them.

**Comment by aaronsteers** [comment]
@test-pr

**Comment by aaronsteers** [comment]
@Suraj-Vishwakarma70 - In order to perform a final review on this, we'll need these two CI tests to be in passing (🟢 green) status:

- `Run Tests / Pytest (Fast)` ([log here](https://github.com/airbytehq/PyAirbyte/actions/runs/9849710999/job/27193820827?pr=285))
- `Run Tests / Pytest (No Creds)` ([log here](https://github.com/airbytehq/PyAirbyte/actions/runs/9849710999/job/27193821146?pr=285))

File: airbyte/sources/base.py [hunk 2]
```diff
@@ -761,3 +823,4 @@ def read(
 __all__ = [
     "Source",
 ]
++
```
Comment (aaronsteers) [inline airbyte/sources/base.py]:
```suggestion

```

**Comment by Suraj-Vishwakarma70** [comment]
@aaronsteers I've fixed some stuff! Please have a look!

File: airbyte/sources/base.py [hunk 3]
```diff
@@ -99,13 +101,25 @@ def set_streams(self, streams: list[str]) -> None:
...
     def select_all_streams(self) -> None:
         """Select all streams.
 
         This is a more streamlined equivalent to:
         > source.select_streams(source.get_available_streams()).
         """
+        if self._config_dict is None:
+            self._to_be_selected_streams = "*"
+            self._log_warning_preselected_stream(self._to_be_selected_streams)
+            return
```
[the remark this answers is not in the document]
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py:125]: We will continue using print ,becuase thats what is being used in PyAirbyte!

File: airbyte/sources/base.py [hunk 4]
```diff
@@ -99,13 +101,25 @@
             stacklevel=2,
         )
         self.select_streams(streams)
+        
+    def _log_warning_preselected_stream(self, streams: str | list[str]) -> None:
+        """Logs a warning message indicating stream selection which are not selected yet"""
+        if streams == "*":
+            print("Warning: Config is not set yet. All streams will be selected after config is set.")
+        else:
+            print(f"Warning: Config is not set yet. {streams} streams will be selected after config is set.")
```
[the remark this answers is not in the document]
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py]: We will continue using print ,becuase thats what is being used in PyAirbyte!

File: airbyte/sources/base.py [hunk 5]
```diff
@@ -116,6 +133,11 @@
 
         Currently, if this is not set, all streams will be read.
         """
+        if self._config_dict is None:
+            self._to_be_selected_streams = streams
+            self._log_warning_preselected_stream(streams)
+            return
```
[the remark this answers is not in the document]
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py:140]: We will continue using print ,becuase thats what is being used in PyAirbyte!

File: airbyte/sources/base.py [hunk 6]
```diff
@@ -100,12 +101,28 @@ def set_streams(self, streams: list[str]) -> None:
...
+    def _log_warning_preselected_stream(self, streams: str | list[str]) -> None:
+        """Logs a warning message indicating stream selection which are not selected yet"""
+        if streams == "*":
+            print(
+                "Warning: Config is not set yet. All streams will be selected after config is set."
+            )
+        else:
+            print(
+                f"Warning: Config is not set yet. {streams} streams will be selected after config is set."
+            )
```
[the remark this answers is not in the document]
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py]: We will continue using print ,becuase thats what is being used in PyAirbyte!

File: airbyte/sources/base.py [hunk 7]
```diff
@@ -81,6 +81,7 @@
         self._discovered_catalog: AirbyteCatalog | None = None
         self._spec: ConnectorSpecification | None = None
         self._selected_stream_names: list[str] = []
+        self._to_be_selected_stream: list[str] | str = []
```
[the remark this answers is not in the document]
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py]: Typos have been fixed with next commit

File: airbyte/sources/base.py [hunk 8]
```diff
@@ -159,6 +181,10 @@
 
         self._config_dict = config
 
+        if self._to_be_selected_stream:
+            self.select_streams(self._to_be_selected_stream)
+            self._to_be_selected_streams = []
+
```
[the remark this answers is not in the document]
Comment (Suraj-Vishwakarma70) [reply airbyte/sources/base.py]: Typos have been fixed in next commit!

File: airbyte/sources/base.py (hunk 6 above)
Comment (aaronsteers) [inline airbyte/sources/base.py]: Line length can't be auto-fixed by the linter...

```suggestion
            print(
                "Warning: Config is not set yet. "
                f"Streams to be selected after config is set: {streams}"
            )
```

**Comment by aaronsteers** [comment]
@Suraj-Vishwakarma70 - I think this is close. If tests pass at this stage, I think we may be ready to merge. Quick note: unfortunately, `/fix-pr` and `/test-pr` will only work for administrators, due to permissions.

**Pull request merged**